<launch>

  <!-- audio_to_spectrogram arguments -->
  <arg name="device" default="hw:0,0" />
  <arg name="n_channel" default="2" />
  <arg name="bitdepth" default="16" />
  <arg name="mic_sampling_rate" default="44100" />
  <arg name="use_rosbag" default="false" />
  <arg name="filename" default="/" />
  <arg name="use_microphone" default="true" />
  <arg name="high_cut_freq" default="8000" />
  <arg name="low_cut_freq" default="1" />
  <arg name="spectrogram_period" default="1" />
  <arg name="pause_rosbag" default="true" />
  <arg name="gui" default="true" />

  <!-- sound_saver arguments -->
  <arg name="save_data_rate" default="5"/>
  <arg name="target_class" default="" />
  <arg name="save_when_sound" default="true"/>

  <!-- sound_detector_volume arguments -->
  <arg name="threshold" default="0.5"/>

  <include file="$(find sound_classification)/launch/audio_to_spectrogram.launch" >
    <arg name="device" value="$(arg device)" />
    <arg name="n_channel" value="$(arg n_channel)" />
    <arg name="bitdepth" value="$(arg bitdepth)" />
    <arg name="mic_sampling_rate" value="$(arg mic_sampling_rate)" />
    <arg name="use_rosbag" value="$(arg use_rosbag)" />
    <arg name="filename" value="$(arg filename)" />
    <arg name="use_microphone" value="$(arg use_microphone)" />
    <arg name="high_cut_freq" value="$(arg high_cut_freq)" />
    <arg name="low_cut_freq" value="$(arg low_cut_freq)" />
    <arg name="spectrogram_period" value="$(arg spectrogram_period)" />
    <arg name="pause_rosbag" value="$(arg pause_rosbag)" />
    <arg name="gui" value="$(arg gui)" />
  </include>

  <!-- Collect spectrogram with sound class, only when the robot is in sound. -->
  <node pkg="sound_classification" type="sound_saver.py" name="sound_saver" output="screen">
    <remap from="~in_sound" to="/sound_detector_volume/in_sound" />
    <remap from="~input" to="/preprocess_gray_image/output_normalized" />
    <remap from="~input_raw" to="/spectrum_to_spectrogram/spectrogram" />
    <rosparam subst_value="true">
      <!-- params below is enabled only when save_data is true -->
      save_data_rate: $(arg save_data_rate) <!-- related to spectrogram_period -->
      target_class: $(arg target_class)
      save_when_sound: $(arg save_when_sound)
      save_raw_spectrogram: true
    </rosparam>
  </node>

  <!-- Detect sound based on pixel value sum (nearly equals to volume) -->
  <node pkg="sound_classification" type="sound_detector_volume.py"
        name="sound_detector_volume" respawn="true" output="screen">
    <remap from="~input" to="/preprocess_gray_image/output" />
    <rosparam subst_value="true">
      power_per_pixel_threshold: $(arg threshold)
      lazy: false
    </rosparam>
  </node>

</launch>
